flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -685,13 +691,24 @@ JobManagerRunner createJobManagerRunner(JobGraph jobGraph, long initializationTi
 
     @Override
     public CompletableFuture<String> triggerSavepoint(
-            final JobID jobId,
+            final AsynchronousJobOperationKey operationKey,
             final String targetDirectory,
             final boolean cancelJob,
             final Time timeout) {
 
-        return performOperationOnJobMasterGateway(
-                jobId, gateway -> gateway.triggerSavepoint(targetDirectory, cancelJob, timeout));
+        try {
+            // Check if a trigger request for this key exists already
+            return savepointOperationsCache.get(operationKey);
+        } catch (UnknownOperationKeyException e) {
```

Comment:
We shouldn't use exceptions to determine expected conditions.

Reply:
True. I guess I'll add some `containsKey` method to the cache